### PR TITLE
WIP: Feature/store encryption

### PR DESCRIPTION
**Pull request by gautier-lefebvre**
- Encrypt / decrypt the state using the user password inside a file in %APP_DATA%
- Handlers for IPC requests in main process

### CODE REVIEW

File: src/main/state/encrypt/index.js [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+const fse = require('fs-extra');
+
+const ALGORITHM = 'aes-256-cbc';
+
+/**
+ * Encrypt the state with the given key.
+ *
+ * @param {Object} state - State tree.
+ * @param {string} key - Key to encrypt, as an utf8 string.
+ * @returns {string} The state encrypted, as an utf8 string.
```
Comment (sylvainar) [inline src/main/state/encrypt/index.js]: Don't we have a jsdoc rule imposing an hyphen between the parameter type and the description ?

Comment (gautier-lefebvre) [reply src/main/state/encrypt/index.js]: Like that ?
```js
/**
 * @returns {string} - The state encrypted, as an utf8 string.
 */
```

I don't think so, only between the param name and its description.

Comment (sylvainar) [reply src/main/state/encrypt/index.js]: Yup, like that. Should we ?

Comment (gautier-lefebvre) [reply src/main/state/encrypt/index.js]: I don't think so 😄

File: src/main/config/constants.js [hunk 2]
```diff
@@ -0,0 +1,8 @@
+const path = require('path');
+
+const { app } = require('electron');
+
+module.exports = {
+  /** Path to the file where the store is saved. */
+  saveFilePath: path.join(app.getPath('userData'), 'store.dvdb'),
```
Comment (maxence-lefebvre) [inline src/main/config/constants.js]: storeFilePath ?

File: src/main/state/encrypt/index.js [hunk 3]
```diff
@@ -0,0 +1,84 @@
+/** From http://lollyrock.com/articles/nodejs-encryption/. */
+
+const crypto = require('crypto');
+
+const fse = require('fs-extra');
+
+const ALGORITHM = 'aes-256-cbc';
```
Comment (maxence-lefebvre) [inline src/main/state/encrypt/index.js]: in constants ?

File: src/main/state/encrypt/index.js [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+ * Decrypt the state with the given key.
+ *
+ * @param {Object} encryptedState - Encrypted state, as an utf8 string.
+ * @param {string} key - Key used to encrypt, as an utf8 string.
+ * @returns {Object} The state tree, decrypted.
+ */
+function decrypt(encryptedState, key) {
+  const decipher = crypto.createDecipher(ALGORITHM, key);
+  const decrypted = decipher.update(encryptedState, 'base64', 'utf8') + decipher.final('utf8');
+  return JSON.parse(decrypted);
```
Comment (maxence-lefebvre) [inline src/main/state/encrypt/index.js]:
```js
try {
  return JSON.parse(decrypted);
catch() { 
  // TODO defineMessage ?
  throw 'corrupted' ;
}
```

File: src/main/state/encrypt/index.js [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+/**
+ * Encrypt the state and write to a file.
+ *
+ * @async
+ * @param {Object} state - State tree.
+ * @param {string} key - Key to encrypt, as an utf8 string.
+ * @param {string} saveFilePath - Path to the file where to write.
+ */
+function encryptToFile(state, key, saveFilePath) {
```
Comment (maxence-lefebvre) [inline src/main/state/encrypt/index.js]: persistEncryptedState ?

File: src/renderer/redux/middlewares/saveState.js [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+        }
+
+        isIdle = true;
+        return null;
+      });
+  }
+
+  return store => next => (action) => {
+    // If the state is being saved, keep the new tree in memory but don't try to save again.
+    if (!isIdle) {
```
Comment (maxence-lefebvre) [inline src/renderer/redux/middlewares/saveState.js]: wouldn't you want to debounce to save at maximum once a second  with the state of the last call ?

**Review by maxence-lefebvre — APPROVED** [review]

**Pull request merged**